### ISSUE TITLE
Pylint false positives on return value from random.sample

### ISSUE THREAD

**Issue by quodlibetor** [PyCQA/pylint#811]
I'm getting these false positives in both python2.7 and python3.4, with pylint 1.5.4.

Given the following program:

``` python
'test'

import random


VALS = [{'a': 1}, {'a': 2}, {'a': 3}]


def not_subscriptable():
    'test'
    subvals = random.sample(VALS, 2)

    return [v['a'] for v in subvals]


def not_an_index():
    'test'
    idxs = random.sample(range(len(VALS)), 2)
    subvals = [VALS[i] for i in idxs]

    return [v['a'] for v in subvals]


print('not_subscriptable', not_subscriptable())
print('not_an_index', not_an_index())
```

I would expect pylint to run successfully, instead I get:

``` shell
$ pylint pylint_subscript.py
No config file found, using default configuration
************* Module pylint_subscript
E: 13,12: Value 'v' is unsubscriptable (unsubscriptable-object)
E: 19, 0: Sequence index is not an int, slice, or instance with __index__ (invalid-sequence-index)

```

Python is obviously fine with it:

```
$ python3 pylint_subscript.py
not_subscriptable [3, 1]
not_an_index [2, 1]
```

```
$ python2 pylint_subscript.py
('not_subscriptable', [3, 2])
('not_an_index', [2, 1])
```

**Comment by PCManticore** [comment]
Thanks, will look into it.

**Comment by moylop260** [comment]
I had the same issue with the following code:
```python
"""docstring"""

def method(vals):
    """docstring"""
    criteria = [False, False]
    # more code ...
    criteria[0] = vals[1]
    # more code ...
    value = criteria[0][1]
    # more code ...
    return value


method(["string1", "string2"])

```
```bash
pylint --version
pylint 2.0.0,
astroid 1.5.0
Python 2.7.10 (default, Oct 23 2015, 19:19:21)
[GCC 4.2.1 Compatible Apple LLVM 7.0.0 (clang-700.0.59.5)]

```

Result:
<img width="879" alt="screen shot 2016-05-06 at 10 33 07 am" src="https://cloud.githubusercontent.com/assets/6644187/15077843/f315d0d8-1375-11e6-9574-3af123b63dd5.png">

**Comment by PCManticore** [comment]
@moylop260 we don't do reassignment evaluation yet, so this might be triggered for now. There are plans to support it, as well as the case for the OP, but I don't know when I'll get to it.

**Comment by andy-maier** [comment]
I also see that false positive, and instead of creating a new issue I decided to provide more data to this issue here. In my case its occurrence depends e.g. on the order of if-then-else branches:

```
#!/usr/bin/env python

# pylint: disable=missing-docstring

from __future__ import print_function

def args1(msg):
    if msg is None:
        return (msg, None)
    else:
        return (msg, ["hello", "world"])

def args2(msg):
    return (msg, ["hello", "world"])

def args3(msg):
    if msg is not None:
        return (msg, ["hello", "world"])
    else:
        return (msg, None)

print(args1("foo")[1][0])
print(args2("foo")[1][0])
print(args3("foo")[1][0])
```

PyLint result:

```
$ pylint -r n pylint_unsubscriptable.py
No config file found, using default configuration
************* Module pylint_unsubscriptable
I:  3, 0: Locally disabling missing-docstring (C0111) (locally-disabled)
E: 22, 6: Value 'args1('foo')[1]' is unsubscriptable (unsubscriptable-object)
```
Note that it raises "unsubscriptable-object" only in case of `args1()` but not for `args2()` or `args3()`.

PyLint versions:
```
pylint 1.5.5,
astroid 1.4.5
Python 2.7.6 (default, Jun 22 2015, 17:58:13)
[GCC 4.8.2]
```

**Comment by jnothman** [comment]
I presume pylint's popularity as an IDE integration is on the rise, as we are starting to receive complaints due to this issue over at scikit-learn. It applies where a function changes its return value type on the basis of a parameter, which is common in a number of scientific python applications (esp. to return a tuple rather than a single value).

@PCManticore: would a suitable workaround for the time being mark a return value as being of unknown type if a function has multiple returns? Where would a contributor implement such a thing?

**Comment by PCManticore** [comment]
@jnothman I added a change to astroid that could potentially solve this issue. If you can test it, let me know if it still reproduces for you.

**Comment by jnothman** [comment]
The issue i was talking about was not in random.sample, but related to the
latter comment about different return types in different conditional
branches. but indeed, I've not tried to ascertain if that issue has been
fixed

**Comment by PCManticore** [comment]
Thank for clarifying that, @jnothman, I think I missed your last part. This change would probably go either in astroid, but I would say it could go as well in pylint. Since astroid is the low level component, we would still want to infer all the potential types (especially in a world where we could understand control flow and attach types to each execution branch), so the onus will then be on pylint to ensure that it is not emitting anything if there are multiple types involved depending on one branch or another. You could take a look in `pylint/checkers/typecheck.py` where we do the type-checking. Also make sure to check `astroid/inference.py` as well as `astroid/protocols.py`. Probably the former would be a place to add a potential inference mechanism for multiple branches. Let me know if this helps, happy to chat more if you need any help.

**Comment by jnothman** [comment]
thanks. my load is a bit heavy right now, but perhaps we'll be able to find
someone to work on it by leaving them breadcrumbs to follow.
